sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala

```diff
@@ -212,6 +212,119 @@ class CollationStringExpressionsSuite
+      SubstringTestCase(
+        "select left(' a世a ' collate " + "utf8_binary" + ", 3)", // scalastyle:ignore
+        "utf8_binary",
+        Row(" a世")), // scalastyle:ignore
+      SubstringTestCase(
+        "select right(' a世a ' collate " + "unicode" + ", 3)", // scalastyle:ignore
+        "unicode",
+        Row("世a ")), // scalastyle:ignore
+      SubstringTestCase(
+        "select left('ÀÃÂĀĂȦÄäåäáâãȻȻȻȻȻǢǼÆ' collate " + "unicode_ci" + ", 3)", // scalastyle:ignore
+        "unicode_ci",
+        Row("ÀÃÂ")), // scalastyle:ignore
+      SubstringTestCase(
+        "select right('ÀÃÂĀĂȦÄäåäáâãȻȻȻȻȻǢǼÆ' collate " + "utf8_binary_lcase" + ", 3)", // scalastyle:ignore
+        "utf8_binary_lcase",
+        Row("ǢǼÆ")), // scalastyle:ignore
+      SubstringTestCase(
+        "select substr('' collate " + "utf8_binary_lcase" + ", 1, 1)",
+        "utf8_binary_lcase",
+        Row("")),
+      SubstringTestCase(
+        "select substr('' collate " + "unicode" + ", 1, 1)",
+        "unicode",
+        Row("")),
+      SubstringTestCase(
+        "select left('' collate " + "utf8_binary" + ", 1)",
+        "utf8_binary",
+        Row("")),
+        // improper values
+      SubstringTestCase(
+        "select left(null collate " + "utf8_binary_lcase" + ", 1)",
+        "utf8_binary_lcase",
+        Row(null)),
+      SubstringTestCase(
+        "select right(null collate " + "unicode" + ", 1)",
+        "unicode",
+        Row(null)),
+      SubstringTestCase(
+        "select substr(null collate " + "utf8_binary" + ", 1)",
+        "utf8_binary",
+        Row(null)),
+      SubstringTestCase(
+        "select substr(null collate " + "unicode_ci" + ", 1, 1)",
+        "unicode_ci",
+        Row(null)),
+      SubstringTestCase(
+        "select left(null collate " + "utf8_binary_lcase" + ", null)",
+        "utf8_binary_lcase",
+        Row(null)),
+      SubstringTestCase(
+        "select right(null collate " + "unicode" + ", null)",
+        "unicode",
+        Row(null)),
+      SubstringTestCase(
+        "select substr(null collate " + "utf8_binary" + ", null)",
+        "utf8_binary",
+        Row(null)),
+      SubstringTestCase(
+        "select substr(null collate " + "unicode_ci" + ", null, null)",
+        "unicode_ci",
+        Row(null)),
+      SubstringTestCase(
+        "select left('ÀÃÂĀĂȦÄäåäáâãȻȻȻȻȻǢǼÆ' collate " + "utf8_binary_lcase" + ", null)", // scalastyle:ignore
+        "utf8_binary_lcase",
+        Row(null)),
+      SubstringTestCase(
+        "select right('ÀÃÂĀĂȦÄäåäáâãȻȻȻȻȻǢǼÆ' collate " + "unicode" + ", null)", // scalastyle:ignore
+        "unicode",
+        Row(null)),
+      SubstringTestCase(
+        "select substr('ÀÃÂĀĂȦÄäåäáâãȻȻȻȻȻǢǼÆ' collate " + "utf8_binary" + ", null)", // scalastyle:ignore
+        "utf8_binary",
+        Row(null)),
+      SubstringTestCase(
+        "select substr('' collate " + "unicode_ci" + ", null, null)",
+        "unicode_ci",
+        Row(null))
```

Comment:
please follow the format of tests in this suite, queries should be constructed from these parameterized cases, otherwise they don't have a point

Reply:
@uros-db I will change it accordingly.  Please advise - Do you want three case classes, or one case class but with a parameter for function name?  If the latter (one case class), how do you want me to handle the third parameter (`len`), which `left` and `right` do not have, and which is optional for `substr`? Maybe with an `Option[String]`?

Is the quantity of tests satisfactory? I got it down from 112 tests to 25 tests. Thus 13 for valid values  and 12 for invalid values. I can get it down to 12 valid test cases and, say, six invalid values if you prefer.

Reply:
It's fine if there's a parameter and it's not used in some cases, I don't think that causes any error

otherwise you can always introduce `LeftRightTestCase`, whatever works

quantity is fine here, what's important is proper coverage without needless repetition, these e2e sql tests are pretty slow, so having hundreds of them for trivial expressions is less than ideal